coderd/workspaceupdates_test.go

```diff
@@ -0,0 +1,366 @@
+					Name:        "agent3",
+					WorkspaceId: ws3IDSlice,
+				},
+			},
+			DeletedWorkspaces: []*proto.Workspace{},
+			DeletedAgents:     []*proto.Agent{},
+		}, update)
+
+		// Update the database
+		db.orderedRows = []database.GetWorkspacesAndAgentsByOwnerIDRow{
+			{
+				ID:         ws1ID,
+				Name:       "ws1",
+				JobStatus:  database.ProvisionerJobStatusRunning,
+				Transition: database.WorkspaceTransitionStart,
+				Agents: []database.AgentIDNamePair{
+					{
+						ID:   agent1ID,
+						Name: "agent1",
+					},
+					{
+						ID:   agent2ID,
+						Name: "agent2",
+					},
+				},
+			},
+			{
+				ID:         ws2ID,
+				Name:       "ws2",
+				JobStatus:  database.ProvisionerJobStatusRunning,
+				Transition: database.WorkspaceTransitionStop,
+			},
+			{
+				ID:         ws4ID,
+				Name:       "ws4",
+				JobStatus:  database.ProvisionerJobStatusRunning,
+				Transition: database.WorkspaceTransitionStart,
+			},
+		}
+		publishWorkspaceEvent(t, ps, ownerID, &wspubsub.WorkspaceEvent{
+			Kind:        wspubsub.WorkspaceEventKindStateChange,
+			WorkspaceID: ws1ID,
+		})
+
+		update = testutil.RequireRecvCtx(ctx, t, sub.Updates())
+		slices.SortFunc(update.UpsertedWorkspaces, func(a, b *proto.Workspace) int {
+			return strings.Compare(a.Name, b.Name)
+		})
+		require.Equal(t, &proto.WorkspaceUpdate{
+			UpsertedWorkspaces: []*proto.Workspace{
+				{
+					// Changed status
+					Id:     ws2IDSlice,
+					Name:   "ws2",
+					Status: proto.Workspace_STOPPING,
+				},
+				{
+					// New workspace
+					Id:     ws4IDSlice,
+					Name:   "ws4",
+					Status: proto.Workspace_STARTING,
+				},
+			},
+			UpsertedAgents: []*proto.Agent{
+				{
+					Id:          agent2IDSlice,
+					Name:        "agent2",
+					WorkspaceId: ws1IDSlice,
+				},
+			},
+			DeletedWorkspaces: []*proto.Workspace{
+				{
+					Id:     ws3IDSlice,
+					Name:   "ws3",
+					Status: proto.Workspace_STOPPED,
+				},
+			},
+			DeletedAgents: []*proto.Agent{
+				{
+					Id:          agent3IDSlice,
+					Name:        "agent3",
+					WorkspaceId: ws3IDSlice,
+				},
+			},
+		}, update)
+	})
+
+	t.Run("Resubscribe", func(t *testing.T) {
+		t.Parallel()
+
+		ctx := testutil.Context(t, testutil.WaitShort)
+
+		db := &mockWorkspaceStore{
+			orderedRows: []database.GetWorkspacesAndAgentsByOwnerIDRow{
+				{
+					ID:         ws1ID,
+					Name:       "ws1",
+					JobStatus:  database.ProvisionerJobStatusRunning,
+					Transition: database.WorkspaceTransitionStart,
+					Agents: []database.AgentIDNamePair{
+						{
+							ID:   agent1ID,
+							Name: "agent1",
+						},
+					},
+				},
+			},
+		}
+
+		ps := &mockPubsub{
+			cbs: map[string]pubsub.ListenerWithErr{},
+		}
+
+		updateProvider := coderd.NewUpdatesProvider(slogtest.Make(t, nil), ps, db, &mockAuthorizer{})
+		t.Cleanup(func() {
+			_ = updateProvider.Close()
+		})
+
+		sub, err := updateProvider.Subscribe(dbauthz.As(ctx, ownerSubject), ownerID)
+		require.NoError(t, err)
```

Comment:
need to call sub.Close, either with `t.Cleanup` or `defer` -- don't mix them.

Check that you are doing this every time you subscribe, not just here.